### PR TITLE
Fix #39338 by making SaltCacheLoader a singleton

### LINKED ISSUE #39338: ZMQ destroys cached connections randomly
### Description of Issue/Question
After some debugging today, I am 100% sure what's still making my deploy slow
(...after fixing lots of other things like loader, slow formulas rendering time etc.)
The time is wasted in fileclient.py:get_file.
Lots of it.

Basically what I added as a workaround for debugging is that at get_file, I return the file if it exists in the cache without checking if it's up-to-date or using hash_and_stat_file.

get_file is used mostly for retrieving the .sls files, but also for files used in file.managed - the .sls part is much bigger slowdown here

The master-side functions are executing in a fraction of a second(in the master log file), so they are not the problem in my opinion.
(_file_hash etc.)

At this point I think zeromq polling or connection caching could be to blame, but I am going to investigate further - if anyone on the core team would have an idea already, that would be great!
The next thing I am going to check is hash_and_stat, but the files are just small formula files that aren't expensive to hash locally.

Other ideas:
1. Add a master function that returns a map of file<->hash.
This would eliminate the need to call master for each requested file.

Without workaround(+ one sample file.managed task from the output, but there are around 80 states run):

```
----------
          ID: nginx_xx_vhost
    Function: file.managed
        Name: /etc/nginx/sites-enabled/xx
      Result: True
     Comment: File /etc/nginx/sites-enabled/xx is in the correct state
     Started: 03:49:19.688364
    Duration: 94.819 ms
     Changes:   

Total run time:    9.573 s
[TRACE   ] _get_event() waited 0 seconds and received nothing
[DEBUG   ] jid 20170212033150376035 found all minions {'xx'}
ERROR: Minions returned with non-zero exit code
salt xx state.highstate -l all  1.10s user 0.08s system 5% cpu 21.912 total
```

With workaround:

```

          ID: nginx_xx_vhost
    Function: file.managed
        Name: /etc/nginx/sites-enabled/xx
      Result: True
     Comment: File /etc/nginx/sites-enabled/xx is in the correct state
     Started: 03:51:30.080166
    Duration: 38.033 ms
     Changes:   


Total run time:    6.351 s
[TRACE   ] _get_event() waited 0 seconds and received nothing
[DEBUG   ] jid 20170212032842315286 found all minions {'xx'}
ERROR: Minions returned with non-zero exit code
salt xx state.highstate -l all  1.04s user 0.07s system 9% cpu 12.247 total
```

### Setup
(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)
Typical minion-master setup.
Latency between minion and master is 10ms ping.

### Steps to Reproduce Issue
(Include debug logs if possible and relevant.)

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)
2016.11.1, python3

### PR DESCRIPTION
**Pull request by ninja-**
### What does this PR do?

This is my proposed fix for #39338 
Basically the problem is that one ZeroMQ connection should be re-used all over the highstate run.
Doing a handshake is expensive and in my setup can take around 50ms.

In my highstate, there are around 50 re-handshakes caused primarly by poorly instantiated SaltCacheLoader.

But it's hard.
It's hard to pass the original client everywhere.
It's hard to make the client a singleton and cache it's instance properly without breaking stuff.
(I tried)

After evaluating all possible solutions, this one is pretty simple and I'd like to get this merged and later we could think of something better.
(on how to have only one instance of FileClient per minion/master)

This PR fixes a second performance problem with jinja loader.
If you for example had 10 SLS files which included nginx/map.jinja, then SaltCacheLoader would ping-pong the master with _find_file each time.
The ``` self.cache ``` variable used in SaltCacheLoader was useless, because it was thrown away after a single template has been loaded.
So now it's going to ping the master for nginx/map.jinja only once per highstate.

Note: this problem doesn't happen when loading .sls files, because they are loaded from a single FileClient.
It's hard to pass original FileClient to rendering pipeline, because render_* functions are given just a string and not the original file or a context to inject.

Maybe we could make FileClient a global variable?
Or inject the FileClient into __opts__?
^ It would be good to inject original FileClient into opts as long as it's thread safe to use it everywhere - is it? I am _pretty sure_ it is

EDIT: I understand that currently this PR breaks the tests for example, so maybe we could try something different.
How about adding some instance caching here https://github.com/saltstack/salt/blob/63c7c794551293fd28695417e8e5bf232184a341/salt/fileclient.py#L45?
### What issues does this PR fix or reference?

#39338

### Previous Behavior

Create new SaltCacheLoader & FileClient for each new rendered template file.

### New Behavior

Make SaltCacheLoader a singleton - one instance for loading pillar files and one for state files.

### Tests written?

Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

**Comment by cachedout** [comment]
It will take some time to look carefully at this. In the meantime, could you please fix these lint errors? Thanks. https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/8614/violations/

File: salt/utils/jinja.py
Comment (gtmanfred) [inline salt/utils/jinja.py]: should this be a `weakref.WeakKeyDictionary`?

Comment (ninja-) [inline salt/utils/jinja.py]: I am pretty sure it would be garbage collected when it's reset to a new {} without need to use WeakKeyDictionary? The key is (true,) or (false,)

**Comment by gtmanfred** [comment]
Yeah, that makes sense, we have just been talking about singletons today
and I was looking at the transport layer singleton and saw it used a
weakkeydict and was wondering if it made sense to use here too.

Thanks,
Daniel

**Comment by ninja-** [comment]
Also, as I mentioned earlier this PR breaks test in a way that's pretty hard to fix...it would be great if someone on the core team could help in fixing it another way, to cache the FileClient maybe at get_file_client level or injecting the instance into __opts__.
(unless you are fine with using this solution temporarily, then let me know and I'll have a look at fixing the tests)

**Comment by ninja-** [comment]
And to be honest the way it's currently done probably breaks threading or at least leaves a race condition if the fileclient is used after fork or from another thread.

**Review by thatch45 — APPROVED** [review]
I am not a big fan of singletons, this is not my go-to solution usually. But I think it works out very well here and I am on board

**Review by terminalmage — APPROVED** [review]
:+1:

**Comment by terminalmage** [comment]
@ninja- can you rebase against develop and force push? The tests have aged out and there are some lint failures, but if I just restart the tests a bunch will fail since there have been changes to the test suite since you opened this PR.

I think this is most likely good to merge but I would like to see the lint status before we get this in.

**Comment by ninja-** [comment]
@terminalmage @thatch45 guys I really appreciate your approval here but I am seriously aware of how much my code here is broken with threads and the like. Can we work on something better together?

My goal is to have one cached FileClient instance per thread(contextmanager would be useful?) and making sure these instances are wiped after process forks(also closing ZeroMQ connections after fork which would normally be garbage collected).

This template caching logic could be left to FileClient at that point(get_file early-returning with caches file_list), and with some help on getting this done I'll finish "Reduce FileClient latency vol2" on top.

**Comment by thatch45** [comment]
I should clarify, when we approve a PR we are saying that we think the code looks good and we don't foresee issues. As always if tests are failing we don't merge!
As for the context manager I am not sure where to point you to start out.

**Comment by cachedout** [comment]
@ninja- What would you like to do on this one? Are you planning to do further work here or are the specific issues you'd like to discuss in order to get this one merged in?

**Comment by ninja-** [comment]
@cachedout as I said I am out of ideas on how to finish it by myself...I would need some help to finish thread-local version of FileClient.

**Comment by cachedout** [comment]
@terminalmage Can you hop in here?

**Comment by ninja-** [comment]
@cachedout @terminalmage as I said I would like some help on how to make the FileClient thread local + make it fork-safe, Then we will be really good on performance because there will be only one handshake and not a spam of handshakes..then I can finish "optimize fileclient for latency vol2"

**Comment by ninja-** [comment]
as for fork maybe this schema(pseudocode)
pid = process.pid()

get_file_client() 
   if process.pid() != pid:
      wipe client and return new
   return cached client

**Comment by terminalmage** [comment]
``get_file_client()`` is run outside of the client instances themselves, so I'm not sure where you'd be caching the fileclient. One potential place is in the process' ``__opts__`` dict, e.g. something like ``__opts__['__fileclient-NNN']`` (where ``NNN`` is the PID). Then you could do something like:

```python
def get_file_client(opts, pillar=False):
    '''
    Read in the ``file_client`` option and return the correct type of file
    server
    '''
    key = '__fileclient-%d' % os.getpid()  # pylint: disable=E1321
    client = opts.get('file_client', 'remote')
    if pillar and client == 'local':
        client = 'pillar'
        key += '-pillar'
    client_class = {
        'remote': RemoteClient,
        'local': FSClient,
        'pillar': LocalClient,
    }.get(client, RemoteClient)
    # Make sure the instance is the same class type
    if key in opts and not isinstance(opts[key], client_class):
        opts.pop(key)
    try:
        return opts[key]
    except KeyError:
        ret = client_class(opts)
        opts[key] = ret
        return ret
```

The ``isinstance`` check there may be overkill since the same PID on the same host shouldn't use different fileclient classes, aside from when the master (or masterless minion) uses the ``FSClient`` for file requests and ``Pillar`` class for Pillar compilation, and those instances are disambiguated by appending ``-pillar`` to the key in my example code above.

@thatch45 thoughts?

**Comment by ninja-** [comment]
I am ok with this as long as this works. Are there no threads involved in salt? Only via multiprocessing aka fork?

**Comment by ninja-** [comment]
And the above does not solve the problem of garbage collecting FileClient of parent after fork so these duped sockets can get disconnected

**Review by thatch45 — APPROVED** [review]

**Comment by thatch45** [comment]
yes @terminalmage mage is correct and salt should avoid threads and just use processes

**Comment by rallytime** [comment]
@ninja- What is the status of this PR? We haven't heard back from you in a while here. Have your questions been answered sufficiently?

**Comment by rallytime** [comment]
Since we haven't heard back in on this in quite some time, I am going to close this.

We of course can open this again when you're ready. Or you can open a new PR if that is easier for you. Thank you!

**Pull request closed without merging**
